### PR TITLE
Initial Conformance Tests for Gateway API

### PR DESCRIPTION
**Pull request by robscott**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR adds a very basic initial framework for conformance tests. Although there is still plenty of room for both improvement and expansion, I want to keep this initial starting point relatively small and straightforward. This PR is ready for full review, I'm especially interested in feedback on the general structure and concepts in this PR though.

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

**Comment by robscott** [comment]
At long last, I think these tests are ready for full review. They line up with the earlier conformance related GEPs and cover the following:

- GatewayClass
  - Status:
    - Accepted
- Gateway:
  - Provisioning (provisioning in advance is also supported)
  - Status:
    - Ready
- HTTPRoute:
  - Matching:
    - Multiple routes
    - Multiple rules
    - Path
    - Hostname
    - Header
  - Status:
    - Parents

**Comment by robscott** [comment]
Adding a hold because this a big PR and want to have at least a couple LGTMs before merging:

/hold

File: conformance/manifests/base.yaml
Comment (hbagdi) [inline conformance/manifests/base.yaml:11]: Having more specific names with some semantic meaning like "shared-infra-gateway" would be better, especially as more and more tests use this.

File: conformance/tests/httproute.go
Comment (hbagdi) [inline conformance/tests/httproute.go:46]: Good starting point. We may need to have a CLI arg for gwAddr. Just leaving a note, no action is necessary yet.

File: conformance/utils/http/http.go
Comment (hbagdi) [inline conformance/utils/http/http.go:50]: Have you considered having retry here in case the resources are being stabilized while this runs?
Or should these be run only once the status of resources have been verified (and thus ensures a deterministic state).

Comment (robscott) [reply conformance/utils/http/http.go:50]: These tests wait for the appropriate ready status to be set on the Route and Gateway. I think that should be interpreted as the Gateway being ready to handle these requests. Although it's possible that some implementations are setting status prior to proxies being completely updated, I think we should try to keep conformance tests as tight as possible initially so that users can trust a ready status across all conformant implementations.

File: conformance/tests/httproute.go
Comment (robscott) [reply conformance/tests/httproute.go:46]: I'd been working with the assumption that Gateway status addresses would always be set on a ready Gateway. Would that not be possible for some of the implementations relying on provisioning Gateways in advance of running these tests?

File: conformance/utils/roundtripper/roundtripper.go
Comment (robscott) [reply conformance/utils/roundtripper/roundtripper.go:127]: I think that's already the case? One is "Sending Request: ...", and this is "Received Response: ...", any changes you'd recommend?

File: conformance/utils/roundtripper/roundtripper.go
Comment (robscott) [reply conformance/utils/roundtripper/roundtripper.go:37]: These are great questions, I struggled with the appropriate interface here, but ended up with something that was pretty similar to what [ingress-controller-conformance implemented](https://github.com/kubernetes-sigs/ingress-controller-conformance/blob/master/test/http/http.go#L39-L66). I think there are 3 main reasons I avoided the net/http interfaces:

1. I want to be able to support more protocols than HTTP. I actually don't know how to test UDP, but TCP could theoretically recycle the same interface.
2. I want tests to be able to specify a request to example.com/foo and know that it will get routed to the appropriate Gateway automatically. That could technically be done by modifying/copying a http request that was passed in, but feels messier. We could also just make each test case specify this as they structure a request.
3. I wanted to use the same echoserver image as ingress-controller-conformance so we didn't have to maintain a separate one for this. That requires a struct that matches the output from that container.

Open to different approaches here, but that's how I landed up with the current approach.

File: conformance/manifests/httproute-cross-namespace.yaml
Comment (robscott) [reply conformance/manifests/httproute-cross-namespace.yaml]: Updated, thanks!

File: conformance/utils/roundtripper/roundtripper.go
Comment (robscott) [reply conformance/utils/roundtripper/roundtripper.go]: Updated, thanks!

File: conformance/utils/suite/suite.go
Comment (robscott) [reply conformance/utils/suite/suite.go]: Updated, thanks!

File: conformance/utils/roundtripper/roundtripper.go
Comment (robscott) [reply conformance/utils/roundtripper/roundtripper.go]: Updated, thanks!

File: conformance/manifests/base.yaml
Comment (robscott) [reply conformance/manifests/base.yaml:19]: I've added some comments at the top of this file describing the key resources, but unfortunately don't have great ideas for how to make the names of the Gateways more descriptive right now. At this point, the only real variation between them is where they accept routes to bind from:

- gateway-same-namespace (only supports route in same ns)
- gateway-all-namespaces (supports routes in all ns)
- gateway-backend-namespaces (supports routes in ns with backend label)

In the future when we add resources with different protocols, multiple listeners, etc, naming may become a bit more clear. Open to ideas though.

File: conformance/tests/httproute.go
Comment (howardjohn) [inline conformance/tests/httproute.go:85]: Header usage doesn't work I think. By the time we get here go will have canonicalized the header name. IE should be checking `Version` (or some other fix)

File: conformance/utils/kubernetes/helpers.go
Comment (howardjohn) [inline conformance/utils/kubernetes/helpers.go:167]: poll every 5s is needlessly slow for a lot of implementations, IMO. I see about 500ms on my env

File: conformance/utils/kubernetes/helpers.go
Comment (howardjohn) [inline conformance/utils/kubernetes/helpers.go:207]: nit: helpful to log what values were actually received vs expected

File: conformance/utils/kubernetes/helpers.go
Comment (howardjohn) [inline conformance/utils/kubernetes/helpers.go:182]: puttting `t.Helper()` may help logs

File: conformance/utils/kubernetes/helpers.go
Comment (howardjohn) [inline conformance/utils/kubernetes/helpers.go:163]: Need this within `PollImmediate` to capture the logs from there, if that was the intent

Comment (howardjohn) [reply conformance/utils/kubernetes/helpers.go:163]: Actually that makes it log in wait.go code ... not sure how to get the call of `HTTPRouteMustHaveParents` in the output

File: conformance/utils/kubernetes/helpers.go
Comment (howardjohn) [inline conformance/utils/kubernetes/helpers.go:206]: `!=` is not effective on pointer types

File: conformance/utils/kubernetes/helpers.go
Comment (robscott) [reply conformance/utils/kubernetes/helpers.go:167]: Good call, updated to 1 second for now, can lower further if needed.

File: conformance/tests/httproute.go
Comment (robscott) [reply conformance/tests/httproute.go:85]: Good catch, updated to check "Version" for now, but will work on a better fix that makes the header check case insensitive.

File: conformance/manifests/base.yaml
Comment (bowei) [reply conformance/manifests/base.yaml:19]: I know it's more verbose, but does it make sense to repeat the YAML across test cases (i.e. make each case hermetic) rather than trying to deduplicate? I know it's kind of a pain, but sharing the setup can result in coupling between test cases and the creation of a complex "base" case that makes it harder to understand the relationship later on.

File: conformance/tests/httproute.go
Comment (bowei) [reply conformance/tests/httproute.go:46]: My recommendation from an organization perspective is to try to avoid nesting by resource type in the test cases, as we will end up having conformance things that are split between gwc, gw, and *route behaviors.

instead put that in the name of the test go file. 
Also, if we could have the tests be encapsulated into a file each makes it easier to understand organizationally:

HTTPRouteSimpleSameNamespace_test.go:

```
func init() {
  ConformanceTests = append(ConformanceTests, &ConformanceTest{ ...
  })
}
```

Comment (robscott) [reply conformance/tests/httproute.go]: @bowei thanks for the suggestion! I've reworked this to use that structure and I think it will make it much easier to add future test cases + understand existing ones.

File: conformance/utils/kubernetes/helpers.go
Comment (robscott) [reply conformance/utils/kubernetes/helpers.go]: Updated, thanks!

File: conformance/utils/kubernetes/helpers.go
Comment (robscott) [reply conformance/utils/kubernetes/helpers.go]: Woops, missed this, thanks!

File: conformance/manifests/base.yaml
Comment (robscott) [reply conformance/manifests/base.yaml]: Yeah it's going to be tough to find the right balance here. For any route to work, it needs the following:

- Namespace
- Gateway
- Service
- Pods

Spinning all of those up and down for every test would quickly become quite time consuming. Alternatively if we were to leave them running indefinitely, we'd need each test case to be sure that it wasn't overlapping with resources defined by other test cases. This would also result in rather verbose yaml definitions for each test case. A final argument in favor of this shared set of base resources is that it enables conformance tests to be meaningful for implementations that are not currently able to support dynamically provisioning Gateways (this generally means there's a 1:1 relationship between Gateway and controller).

With all that said, more isolated tests would enable us to run tests in parallel which could significantly speed up a test run when there were a large number of tests in play. It would also be easier to understand each individual test case than our current approach that requires an understanding of the base resources.

I think we'll inevitably want to move towards a test suite that has more isolated test definitions, but fortunately that should be a fairly straightforward transition from this current state. I think that the current proposal is an acceptable starting point given the constraints above, but that we should try to avoid expanding the base set of resources, and in the future add test cases that are not entirely dependent on them.

File: conformance/utils/http/http.go
Comment (hbagdi) [reply conformance/utils/http/http.go:50]: Okay. That's fine. We can get feedback and tweak this if needed.

File: conformance/tests/httproute-cross-namespace.go
Comment (hbagdi) [inline conformance/tests/httproute-cross-namespace.go:2]: Can you keep a single test case and add the remaining ones in a separate PR? That would minimize the likelihood of shipping buggy tests. This PR is big as is.

File: conformance/tests/httproute-cross-namespace.go
Comment (hbagdi) [inline conformance/tests/httproute-cross-namespace.go:39]: If the description tells me that it is a negative or positive test, it is better - code and words must match up. Else, the reader has to read the code to figure out the purpose of the test.
```suggestion
	Description: "A single HTTPRoute in the gateway-conformance-web-backend namespace attaches to Gateway in other namespace",
```

File: conformance/tests/httproute-cross-namespace.yaml
Comment (hbagdi) [inline conformance/tests/httproute-cross-namespace.yaml:4]:
```suggestion
  name: web-backend
```

We should avoid putting namespace/name/types in the Name or Namespace keys. That makes the whole thing much harder to parse and often unnecessarily.

**Review by shaneutt — APPROVED** [review]
Once other people's comments are resolved this looks like a great place to start for me. We'll be adding these to our test suite soon, the only thing I was really concerned about going into this was making sure there wasn't any friction for testing "unmanaged mode" (existing gateways) but I don't think what's here will cause any problem with that.

File: conformance/tests/httproute-cross-namespace.go
Comment (robscott) [reply conformance/tests/httproute-cross-namespace.go]: Good call, I've updated most test descriptions, hopefully they're clearer now.

File: conformance/tests/httproute-cross-namespace.yaml
Comment (robscott) [reply conformance/tests/httproute-cross-namespace.yaml]: I definitely agree with the idea of removing type prefixes like `httproute-`, I've done that. I'm less sold on giving the routes more generic names. I think names can be used to provide context on what a Gateway/HTTPRoute is for. For example, a Gateway with a name `same-namespace-only` could make it easier to read and understand a test than a Gateway with a more generic name like `app-lb`.

File: conformance/tests/httproute-cross-namespace.go
Comment (robscott) [reply conformance/tests/httproute-cross-namespace.go]: Good call, I left the simplest test in and have pulled the others out for a follow up PR.

File: conformance/utils/roundtripper/roundtripper.go
Comment (hbagdi) [inline conformance/utils/roundtripper/roundtripper.go:46]: Where is this being used?

File: conformance/utils/roundtripper/roundtripper.go
Comment (hbagdi) [inline conformance/utils/roundtripper/roundtripper.go:50]: copy pasta error

File: conformance/utils/roundtripper/roundtripper.go
Comment (hbagdi) [inline conformance/utils/roundtripper/roundtripper.go:44]: Use string for Protocol/Method - simpler API. This doesn't seem to have an any advantage.

File: conformance/utils/roundtripper/roundtripper.go
Comment (hbagdi) [inline conformance/utils/roundtripper/roundtripper.go:2]: I'd recommend starting with a single `conformance/utils/http` package which includes this file.
Any reason to have so manage packages upfront?

File: conformance/utils/http/http.go
Comment (hbagdi) [inline conformance/utils/http/http.go:48]: Is this function actually used anywhere?

File: conformance/utils/suite/suite.go
Comment (hbagdi) [inline conformance/utils/suite/suite.go:23]: Unwanted line jump.

File: conformance/utils/kubernetes/apply.go
Comment (hbagdi) [inline conformance/utils/kubernetes/apply.go:68]: Please replace all contexts with a 10 or 15-second timeout to avoid long-hanging tests. We can fine-tune the timeout later on.
A better (more tedious) approach would be to have a context being propagated up from the top. Either is fine, please avoid requests without timeout.

File: conformance/utils/kubernetes/apply.go
Comment (hbagdi) [inline conformance/utils/kubernetes/apply.go:58]:
```suggestion
			require.NoErrorf(t, err, "error setting field `spec.GatewayClass` on a Gateway resource")
```

File: conformance/utils/kubernetes/apply.go
Comment (hbagdi) [inline conformance/utils/kubernetes/apply.go:67]: Add a similar log line before c.Get() on line 62.

**Review by hbagdi — COMMENTED** [review]
I didn't look at all the code but I think we can iterate here.
I'm comfortable merging.

File: conformance/utils/suite/suite.go
Comment (robscott) [reply conformance/utils/suite/suite.go:23]: This looks strange but was intentional. I'm trying to group imports into three separate categories: 

- stdlib
- other
- local

A bit more explanation of the structure here https://github.com/golang/tools/commit/ed69e84b1518b5857a9f4e01d1f9cefdcc45246e. And an example of us doing similar grouping elsewhere in the project: https://github.com/kubernetes-sigs/gateway-api/blob/master/cmd/admission/main.go.

File: conformance/utils/roundtripper/roundtripper.go
Comment (robscott) [reply conformance/utils/roundtripper/roundtripper.go:2]: Since this is an interface that implementations can override, I've wanted to keep it as generic and possible and leave the door open for at least TCP. That's a ways out though, and I'm not really sure how we'd structure that. The idea was to separate the interface + default implementation from HTTP specific helper funcs.

File: conformance/utils/kubernetes/apply.go
Comment (robscott) [reply conformance/utils/kubernetes/apply.go:67]: I'm a little hesitant to do that because our test output is already fairly verbose here. Here's the output I'm currently getting:

```
    suite.go:56: Test Setup: Ensuring GatewayClass has been accepted
    suite.go:59: Test Setup: Applying base manifests
    apply.go:81: Updating gateway-conformance-infra Namespace
    apply.go:81: Updating same-namespace Gateway
    apply.go:81: Updating all-namespaces Gateway
    apply.go:81: Updating backend-namespaces Gateway
    apply.go:81: Updating infra-backend-v1 Service
    apply.go:81: Updating infra-backend-v1 Deployment
    apply.go:81: Updating infra-backend-v2 Service
    apply.go:81: Updating infra-backend-v2 Deployment
    apply.go:81: Updating gateway-conformance-app-backend Namespace
    apply.go:81: Updating app-backend-v1 Service
    apply.go:81: Updating app-backend-v1 Deployment
    apply.go:81: Updating app-backend-v2 Service
    apply.go:81: Updating app-backend-v2 Deployment
    apply.go:81: Updating gateway-conformance-web-backend Namespace
    apply.go:81: Updating web-backend Service
    apply.go:81: Updating web-backend Deployment
    suite.go:62: Test Setup: Ensuring Gateways and Pods from base manifests are ready
```

Sometimes there's a mix of `Updating` and `Creating` depending on if I've enabled cleanup or added new resources to the base manifests. In either case, adding a `Getting` before each `Updating` or `Creating` line may add more confusion or noise than it's worth. I think it could be helpful to log before the first API call we're doing in case of a timeout or other connection issue, but that's actually happening earlier in the test flow. Do you think there's a better way to log this output as a whole?

File: conformance/utils/kubernetes/apply.go
Comment (robscott) [reply conformance/utils/kubernetes/apply.go]: Good call, I've updated to use context more intentionally throughout, but will work on a larger follow up in the next PR to see where it makes sense to propagate context further.

File: conformance/utils/kubernetes/apply.go
Comment (robscott) [reply conformance/utils/kubernetes/apply.go]: Updated, thanks!

File: conformance/utils/http/http.go
Comment (robscott) [reply conformance/utils/http/http.go]: It is used quite heavily by the tests I pulled out to #994, have pulled out this function to add back with that PR.

File: conformance/utils/roundtripper/roundtripper.go
Comment (robscott) [reply conformance/utils/roundtripper/roundtripper.go]: Updated, thanks!

File: conformance/utils/roundtripper/roundtripper.go
Comment (robscott) [reply conformance/utils/roundtripper/roundtripper.go:47]: Updated, thanks!

File: conformance/utils/roundtripper/roundtripper.go
Comment (robscott) [reply conformance/utils/roundtripper/roundtripper.go]: Woops, removed, thanks!

File: conformance/utils/suite/suite.go
Comment (robscott) [reply conformance/utils/suite/suite.go:57]: I tried `t.Run` first, but that didn't feel right because these items are not really tests as much as they are setup steps. (The GatewayClass status == accepted is relatively close to a test, but the rest are not). Instead I've added logging before each step, which does seem to make the output much clearer here.

**Comment by robscott** [comment]
Thanks for all the great feedback on this @hbagdi, @howardjohn, @jpeach, and anyone else who's commented. I think I've resolved everything and this should be good to go now, PTAL.

File: conformance/utils/suite/suite.go
Comment (hbagdi) [reply conformance/utils/suite/suite.go:23]: Ah, TIL

File: conformance/utils/suite/suite.go
Comment (hbagdi) [inline conformance/utils/suite/suite.go:40]: Since new arguments may be added to this function, replacing the arguments of this function with a "SuiteOpts" struct could be better.

Comment (robscott) [reply conformance/utils/suite/suite.go]: Updated, thanks!

**Comment by hbagdi** [comment]
@robscott Linter unhappy.

**Comment by robscott** [comment]
Woops, fixed the failing test, thanks for catching that @hbagdi! Removing the hold so next LGTM will result in a merge.

/hold cancel

**Comment by hbagdi** [comment]
/lgtm
Thank you for all your hard work on this, Rob! This one took a lot of effort over the past year.

**Pull request merged**
